### About the data set 

The dataset contains Text reviews. The aim of this assignment is to classsify the review texts into two categories. 

Attribute information: 
 
Dataset has two columns:
Column2: Review Text
Column1: Two Classes. Positive and negative.

In [1]:
import numpy as np 
import pandas as pd
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [2]:
df=pd.read_csv("sentiment_analysis.csv",encoding='latin-1')

In [3]:
df.head(10)

Sentiment                                             Review
0  negative  The international electronic industry company ...
1  positive  With the new production plant the company woul...
2  positive  According to the company 's updated strategy f...
3  positive  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...
4  positive  For the last quarter of 2010 , Componenta 's n...
5  positive  In the third quarter of 2010 , net sales incre...
6  positive  Operating profit rose to EUR 13.1 mn from EUR ...
7  positive  Operating profit totalled EUR 21.1 mn , up fro...
8  positive  TeliaSonera TLSN said the offer is in line wit...
9  positive  STORA ENSO , NORSKE SKOG , M-REAL , UPM-KYMMEN...

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1967 entries, 0 to 1966
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Sentiment  1967 non-null   object
 1   Review     1967 non-null   object
dtypes: object(2)
memory usage: 30.9+ KB


In [5]:
df.shape

(1967, 2)

 <font color='brown'> The dataset contains the business and financial aspects of several companies, providing information on their performance, plans, and market activities. There are 1967 entries and each one is  classified into positive or negative based on their reviews.</font>

In [6]:
df['Sentiment']=df['Sentiment'].replace({'negative':'0','positive':'1'})
df.head(5)

Sentiment                                             Review
0         0  The international electronic industry company ...
1         1  With the new production plant the company woul...
2         1  According to the company 's updated strategy f...
3         1  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...
4         1  For the last quarter of 2010 , Componenta 's n...

In [7]:
from bs4 import BeautifulSoup
import re

In [8]:
def remove_html_tags(text):
    soup = BeautifulSoup(text, 'html.parser')
    return soup.get_text()

In [9]:
def remove_special_chars(text):
    return re.sub(r'[^a-zA-Z\s]', '', text)

In [10]:
df['Review'] = df['Review'].apply(remove_html_tags)
df['Review'] = df['Review'].apply(remove_special_chars)
df.head(5)

C:\Users\Priyadarsshini S\AppData\Local\Temp\ipykernel_85120\2125600460.py:2: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  soup = BeautifulSoup(text, 'html.parser')


Sentiment                                             Review
0         0  The international electronic industry company ...
1         1  With the new production plant the company woul...
2         1  According to the company s updated strategy fo...
3         1  FINANCING OF ASPOCOMP S GROWTH Aspocomp is agg...
4         1  For the last quarter of   Componenta s net sal...

 <font color='brown'> All special characters, numbers and html tags are removed from the dataset.</font>

In [11]:
df['Review'] = df['Review'].str.lower()
df.head(5)

Sentiment                                             Review
0         0  the international electronic industry company ...
1         1  with the new production plant the company woul...
2         1  according to the company s updated strategy fo...
3         1  financing of aspocomp s growth aspocomp is agg...
4         1  for the last quarter of   componenta s net sal...

 <font color='brown'> Entire text data is converted to lowercase.</font>

In [12]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to C:\Users\Priyadarsshini
[nltk_data]     S\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [13]:
stp_wds = stopwords.words('english')
stp_wds.remove('not')
print(stp_wds)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'only', 'own', 'same', 'so', 'than', 'too', '

In [14]:
def remove_stopwords(text):
    words = text.split()  # Split the text into words
    filtered_words = [i for i in words if i not in set(stp_wds)]
    return ' '.join(filtered_words)  # Join the filtered words back into a sentence

In [15]:
df['Review'] = df['Review'].apply(remove_stopwords)
df.head(5)

Sentiment                                             Review
0         0  international electronic industry company elco...
1         1  new production plant company would increase ca...
2         1  according company updated strategy years baswa...
3         1  financing aspocomp growth aspocomp aggressivel...
4         1  last quarter componenta net sales doubled eurm...

 <font color='brown'> Stop words like articles, pronouns, common verb, prepositions, conjunctions are removed.</font>

In [16]:
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()

In [17]:
def stem_text(text):
    words = text.split()  # Tokenize the text into words
    stemmed_words = [ps.stem(i) for i in words]  # Stem each word
    return ' '.join(stemmed_words)  # Join the stemmed words back into a sentence

In [18]:
df['Review'] = df['Review'].apply(stem_text)
df.head(10)

Sentiment                                             Review
0         0  intern electron industri compani elcoteq laid ...
1         1  new product plant compani would increas capac ...
2         1  accord compani updat strategi year baswar targ...
3         1  financ aspocomp growth aspocomp aggress pursu ...
4         1  last quarter componenta net sale doubl eurm eu...
5         1  third quarter net sale increas eur mn oper pro...
6         1  oper profit rose eur mn eur mn correspond peri...
7         1    oper profit total eur mn eur mn repres net sale
8         1  teliasonera tlsn said offer line strategi incr...
9         1  stora enso norsk skog mreal upmkymmen credit s...

 <font color='brown'> Stemming is the process of removing inflections from the word and getting the base or root word.</font>

<font color='brown'>"Bag of Words" (BoW) is a common technique in natural language processing (NLP) and text analysis. It converts a collection of text documents into numerical features that can be used to give inputs into an ML model.</font>

In [19]:
corpus = []
for i in range(0, len(df)):
    corpus.append(df['Review'][i])
corpus    

['intern electron industri compani elcoteq laid ten employe tallinn facil contrari earlier layoff compani contract rank offic worker daili postime report',
 'new product plant compani would increas capac meet expect increas demand would improv use raw materi therefor increas product profit',
 'accord compani updat strategi year baswar target longterm net sale growth rang oper profit margin net sale',
 'financ aspocomp growth aspocomp aggress pursu growth strategi increasingli focus technolog demand hdi print circuit board pcb',
 'last quarter componenta net sale doubl eurm eurm period year earlier move zero pretax profit pretax loss eurm',
 'third quarter net sale increas eur mn oper profit eur mn',
 'oper profit rose eur mn eur mn correspond period repres net sale',
 'oper profit total eur mn eur mn repres net sale',
 'teliasonera tlsn said offer line strategi increas ownership core busi hold would strengthen eesti telekom offer custom',
 'stora enso norsk skog mreal upmkymmen credit 

In [20]:
# Creating the Bag of Words model
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000)
X = cv.fit_transform(corpus).toarray()
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [21]:
samp=df.iloc[:]
samp.head(5)

Sentiment                                             Review
0         0  intern electron industri compani elcoteq laid ...
1         1  new product plant compani would increas capac ...
2         1  accord compani updat strategi year baswar targ...
3         1  financ aspocomp growth aspocomp aggress pursu ...
4         1  last quarter componenta net sale doubl eurm eu...

In [22]:
Y= df.iloc[:,0]
Y.head(5)

0    0
1    1
2    1
3    1
4    1
Name: Sentiment, dtype: object

In [23]:
Y=df.iloc[:,0].values
Y

array(['0', '1', '1', ..., '0', '0', '0'], dtype=object)

In [24]:
# Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.30, random_state = 40)

4.3 - Define the Classification model (any one like NB, SVM or Random forest) train the model

In [25]:
# Training model using Naive bayes classifier
from sklearn.naive_bayes import MultinomialNB
classifier = MultinomialNB().fit(X_train, y_train)
y_pred=classifier.predict(X_test)

In [26]:
y_pred

array(['1', '1', '1', '1', '0', '1', '1', '1', '1', '1', '0', '0', '1',
       '1', '0', '1', '0', '1', '0', '1', '0', '1', '1', '0', '1', '1',
       '0', '1', '1', '1', '1', '1', '1', '1', '1', '0', '1', '1', '1',
       '1', '1', '1', '1', '0', '0', '1', '0', '1', '1', '1', '1', '1',
       '1', '0', '1', '1', '0', '0', '1', '0', '0', '1', '0', '1', '1',
       '0', '1', '0', '1', '0', '0', '0', '1', '1', '1', '0', '1', '0',
       '1', '0', '0', '1', '1', '1', '1', '1', '0', '1', '1', '1', '1',
       '1', '0', '1', '1', '1', '1', '1', '1', '1', '1', '1', '0', '1',
       '1', '1', '0', '1', '1', '0', '0', '1', '1', '1', '1', '0', '0',
       '1', '1', '1', '1', '1', '0', '1', '1', '1', '0', '0', '0', '1',
       '0', '1', '1', '0', '0', '0', '1', '1', '0', '1', '1', '1', '1',
       '1', '1', '1', '1', '0', '0', '0', '1', '0', '1', '1', '1', '1',
       '0', '0', '1', '1', '1', '0', '1', '0', '0', '1', '1', '1', '1',
       '1', '0', '1', '1', '1', '0', '1', '1', '1', '1', '0', '1

In [27]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print("Confusion_matrix:\n", cm)

Confusion_matrix:
 [[127  47]
 [ 50 367]]


In [28]:
from sklearn.metrics import classification_report
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.72      0.73      0.72       174
           1       0.89      0.88      0.88       417

    accuracy                           0.84       591
   macro avg       0.80      0.80      0.80       591
weighted avg       0.84      0.84      0.84       591



In [29]:
print("Accuracy_score:",accuracy_score(y_test, y_pred))

Accuracy_score: 0.8358714043993232


 <font color='brown'>The accuracy score of the ML model constructed by using Multinomial Naive Bayes classifier and CountVectorizer is 0.84 </font>

In [30]:
test_review1 = "In the last quarter of 2010, Espresso financials witnessed a remarkable performance."
tst_review_bow1 = cv.transform([test_review1])
tst_pred1 = classifier.predict(tst_review_bow1)

if tst_pred1 == ['1']:
    print("Predicted class of test review is :",tst_pred1, "--> POSITIVE")
else:
    print("Predicted class of test review is :",tst_pred1, "--> NEGATIVE")


Predicted class of test review is : ['1'] --> POSITIVE


In [31]:
test_review2 = "Brownwich Newyork recently laid off several employees at its 4 sites in the city"
tst_review_bow2 = cv.transform([test_review2])
tst_pred2 = classifier.predict(tst_review_bow2)

if tst_pred2 == ['1']:
    print("Predicted class of test review is :",tst_pred2, "--> POSITIVE")
else:
    print("Predicted class of test review is :",tst_pred2, "--> NEGATIVE")

Predicted class of test review is : ['0'] --> NEGATIVE


 <font color='green'> The test-review results are predicted correctly as expected. The output of review1 is  "positive" as the tone of review is clearly talking about the good performance of the company. The output of review2 is "negative" as its says company has laid down employees which might be due to poor profits in that fiscal year.</font>

In [32]:
# Creating the TFIDF model
from sklearn.feature_extraction.text import TfidfVectorizer
tv = TfidfVectorizer(max_features=4000)
X_tf = tv.fit_transform(corpus).toarray()
X_tf

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [33]:
# Train Test Split
from sklearn.model_selection import train_test_split
Xtf_train, Xtf_test, ytf_train, ytf_test = train_test_split(X_tf, Y, test_size = 0.30, random_state = 40)

In [34]:
from sklearn.naive_bayes import MultinomialNB
tf_classifier = MultinomialNB().fit(Xtf_train, ytf_train)
ytf_pred=tf_classifier.predict(Xtf_test)

In [35]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm_tf = confusion_matrix(ytf_test, ytf_pred)
print("Confusion_matrix:\n", cm_tf)

Confusion_matrix:
 [[ 38 136]
 [  3 414]]


In [36]:
from sklearn.metrics import classification_report
report_tf = classification_report(ytf_test, ytf_pred)
print(report_tf)

              precision    recall  f1-score   support

           0       0.93      0.22      0.35       174
           1       0.75      0.99      0.86       417

    accuracy                           0.76       591
   macro avg       0.84      0.61      0.60       591
weighted avg       0.80      0.76      0.71       591



In [37]:
print("Accuracy_score:",accuracy_score(ytf_test, ytf_pred))

Accuracy_score: 0.7648054145516074


 <font color='brown'>The accuracy score of the ML model constructed by using Multinomial Naive Bayes classifier and TF-IDF Vectorizer is 0.765 </font>

Testing own review

In [38]:
test_review3 = "In the last quarter of 2010, Espresso financials witnessed a remarkable performance."
tst_review_bow3 = cv.transform([test_review3])
tst_pred3 = tf_classifier.predict(tst_review_bow3)

if tst_pred3 == ['1']:
    print("Predicted class of test review is :",tst_pred3, "--> POSITIVE")
else:
    print("Predicted class of test review is :",tst_pred3, "--> NEGATIVE")


Predicted class of test review is : ['1'] --> POSITIVE


In [39]:
test_review4 = "Brownwich Newyork recently laid off several employees at its 4 sites in the city"
tst_review_bow4 = cv.transform([test_review4])
tst_pred4 = tf_classifier.predict(tst_review_bow4)

if tst_pred4 == ['1']:
    print("Predicted class of test review is :",tst_pred4, "--> POSITIVE")
else:
    print("Predicted class of test review is :",tst_pred4, "--> NEGATIVE")

Predicted class of test review is : ['0'] --> NEGATIVE


 <font color='brown'> The test-review results are predicted correctly as expected. 
  
  <font color='brown'>The output of review1 is  "positive" as the tone of review is clearly talking about the good performance of the company.
    
 <font color='brown'>The output of review2 is "negative" as its says company has laid down employees which might be due to poor profits in that fiscal year.</font>